openmc/deplete/flux_operator.py

```diff
@@ -0,0 +1,594 @@
+        n_react : int
+            Number of reactions tracked by :class:`openmc.deplete.Operator`
+        op : openmc.deplete.FluxDepletionOperator
+            Reference to the object encapsulate _FluxDepletionRateHelper.
+            We pass this so we don't have to duplicate the ``number`` object.
+
+
+        Attributes
+        ----------
+        nuc_ind_map : dict of int to str
+            Dictionary mapping the nuclide index to nuclide name
+        rxn_ind_map : dict of int to str
+            Dictionary mapping reaction index to reaction name
+
+        """
+
+        def __init__(self, n_nuc, n_react, op):
+            super().__init__(n_nuc, n_react)
+            rates = op.reaction_rates
+
+            self.nuc_ind_map = {ind: nuc for nuc, ind in rates.index_nuc.items()}
+            self.rxn_ind_map = {ind: rxn for rxn, ind in rates.index_rx.items()}
+            self._op = op
+
+        def generate_tallies(self, materials, scores):
+            """Unused in this case"""
+            pass
+
+        def get_material_rates(self, mat_id, nuc_index, react_index):
+            """Return 2D array of [nuclide, reaction] reaction rates
+
+            Parameters
+            ----------
+            mat_id : int
+                Unique ID for the requested material
+            nuc_index : list of str
+                Ordering of desired nuclides
+            react_index : list of str
+                Ordering of reactions
+            """
+            self._results_cache.fill(0.0)
+
+            volume = self._op.number.get_mat_volume(mat_id)
+            for i_nuc, i_react in product(nuc_index, react_index):
+                nuc = self.nuc_ind_map[i_nuc]
+                rxn = self.rxn_ind_map[i_react]
+                density = self._op.number.get_atom_density(mat_id, nuc)
+                self._results_cache[i_nuc,
+                                    i_react] = self._op.cross_sections[rxn][nuc] * density * volume
+
+            return self._results_cache
+
+    def _get_helper_classes(self, helper_kwargs):
+        """Get helper classes for calculating reation rates and fission yields
+
+        Parameters
+        ----------
+        helper_kwargs : dict
+            Keyword arguments for helper classes
+
+        """
+
+        normalization_mode = helper_kwargs['normalization_mode']
+        fission_yield_opts = helper_kwargs.get('fission_yield_opts', {})
+
+        self._rate_helper = self._FluxDepletionRateHelper(
+            self.reaction_rates.n_nuc, self.reaction_rates.n_react, self)
+        if normalization_mode == "constant-power":
+            self._normalization_helper = self._FluxDepletionNormalizationHelper(self)
+        else:
+            self._normalization_helper = SourceRateHelper()
+
+        # Select and create fission yield helper
+        fission_helper = ConstantFissionYieldHelper
+        if fission_yield_opts is None:
+            fission_yield_opts = {}
```

Comment:
This suggestion won't work with the current structure of the code, which is why I didn't incorporate it from your previous round of review, but perhaps I should have said something about this.

In `__init__`, we have
```python
def __init__(self,
                  materials,
                  micro_xs,
                  chain_file,
                  keff=None,
                  normalization_mode='constant-flux',
                  fission_q=None,
                  prev_results=None,
                  reduce_chain=False,
                  reduce_chain_level=None,
                  fission_yield_opts=None):
         .
         .
         .
         helper_kwargs = {'normalization_mode': normalization_mode,
                          'fission_yield_opts': fission_yield_opts}
         .
         .
         .
``` 

The default value of `fission_yield_opts` is `None`, and since it is always passed to `helper_kwargs`, the `helper_kwargs.get('fission_yield_opts', {})` fuction inside of `_get_helper_classes` will always return the value of `fission_yield_opts` stored in `helper_kwargs`, and never return the alternate value given (in this case, an empty `dict`). So we need the `if` statement on 373 to avoid an error from trying to pass a `None` value to `fission_helper`.

If you can think of a good way to restructure this so we don't need the `if` statement, please make a suggestion!

Reply:
Ah, right! In that case, let's just move the `if` block further up, i.e.
```Python
if fission_yield_opts is None:
    fission_yield_opts = {}
helper_kwargs = ...
```